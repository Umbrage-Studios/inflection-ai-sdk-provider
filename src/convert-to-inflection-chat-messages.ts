import {
  LanguageModelV1Prompt,
  UnsupportedFunctionalityError,
} from "@ai-sdk/provider";
import { InflectionChatModelId } from "./inflection-chat-settings";

/**
 * Represents a message in the Inflection AI format
 */
export interface InflectionMessage {
  /** The type of the message sender */
  type: "Human" | "AI" | "Instruction" | "Tool";
  /** The text content of the message */
  text: string;
  /** Optional timestamp in seconds since epoch */
  ts?: number;
  /** Tool calls if present */
  tool_calls?: Array<{
    id: string;
    type: "function";
    function: {
      name: string;
      arguments: string;
    };
  }>;
  /** Tool call ID for tool results */
  tool_call_id?: string;
}

export type InflectionContext = InflectionMessage[];

/**
 * Converts a Language Model prompt to Inflection AI's context format
 * @param prompt The input prompt in the Language Model format
 * @param modelId The Inflection model ID being used
 * @returns An array of messages in Inflection AI's format
 * @throws {UnsupportedFunctionalityError} If the prompt contains unsupported content types
 */
export function convertToInflectionChatMessages(
  prompt: LanguageModelV1Prompt,
  modelId: InflectionChatModelId = "inflection_3_pi"
): InflectionContext {
  const context: InflectionContext = [];

  for (const { role, content } of prompt) {
    let text = "";
    const toolCalls: InflectionMessage["tool_calls"] = [];

    if (typeof content === "string") {
      text = content;
    } else {
      // Combine all text parts into a single string
      for (const part of content) {
        switch (part.type) {
          case "text": {
            text += part.text;
            break;
          }
          case "tool-call": {
            if (modelId !== "inflection_3_with_tools") {
              throw new UnsupportedFunctionalityError({
                functionality:
                  "Tool calls are only supported with the inflection_3_with_tools model",
              });
            }
            toolCalls.push({
              id: part.toolCallId,
              type: "function",
              function: {
                name: part.toolName,
                arguments: JSON.stringify(part.args),
              },
            });
            break;
          }
          case "tool-result": {
            if (modelId !== "inflection_3_with_tools") {
              throw new UnsupportedFunctionalityError({
                functionality:
                  "Tool results are only supported with the inflection_3_with_tools model",
              });
            }
            text = JSON.stringify(part.result);
            break;
          }
          case "image": {
            throw new UnsupportedFunctionalityError({
              functionality:
                "Image content parts are not supported by Inflection AI at this time",
            });
          }
          case "file": {
            throw new UnsupportedFunctionalityError({
              functionality:
                "File content parts are not supported by Inflection AI at this time",
            });
          }
          default: {
            const _exhaustiveCheck: never = part;
            throw new Error(
              `Unsupported content part type: ${_exhaustiveCheck}`
            );
          }
        }
      }
    }

    // Skip empty messages unless they have tool calls
    if (!text && !toolCalls.length) {
      continue;
    }

    // Map the role to Inflection's expected type
    switch (role) {
      case "user": {
        context.push({ type: "Human", text });
        break;
      }
      case "assistant": {
        context.push({
          type: "AI",
          text,
          ...(modelId === "inflection_3_with_tools" && toolCalls.length > 0
            ? { tool_calls: toolCalls }
            : {}),
        });
        break;
      }
      case "system": {
        context.push({ type: "Instruction", text });
        break;
      }
      case "tool": {
        // Only add tool messages if they have content
        if (text) {
          const toolCallId =
            typeof content === "string"
              ? undefined
              : content.find((part) => part.type === "tool-result")?.toolCallId;
          context.push({
            type: "Tool",
            text,
            ...(toolCallId ? { tool_call_id: toolCallId } : {}),
          });
        }
        break;
      }
      default: {
        const _exhaustiveCheck: never = role;
        throw new Error(`Unsupported role: ${_exhaustiveCheck}`);
      }
    }
  }

  return context;
}
