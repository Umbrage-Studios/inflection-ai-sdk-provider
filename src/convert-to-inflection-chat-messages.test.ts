import { expect, describe, it } from "vitest";
import {
  UnsupportedFunctionalityError,
  LanguageModelV1Message,
} from "@ai-sdk/provider";
import { convertToInflectionChatMessages } from "./convert-to-inflection-chat-messages";

describe("user messages", () => {
  it("should convert basic user messages", () => {
    const result = convertToInflectionChatMessages([
      {
        role: "user",
        content: [{ type: "text", text: "Hello" }],
      },
    ]);

    expect(result).toMatchSnapshot();
  });

  it("should combine multiple text parts in user messages", () => {
    const result = convertToInflectionChatMessages([
      {
        role: "user",
        content: [
          { type: "text", text: "Hello " },
          { type: "text", text: "world!" },
        ],
      },
    ]);

    expect(result).toMatchSnapshot();
  });

  it("should throw error for image content in user messages", () => {
    const messages: LanguageModelV1Message[] = [
      {
        role: "user",
        content: [
          {
            type: "image",
            image: new Uint8Array([1, 2, 3]),
            mimeType: "image/jpeg",
          },
        ],
      },
    ];

    expect(() => convertToInflectionChatMessages(messages)).toThrow(
      UnsupportedFunctionalityError
    );
  });

  it("should skip empty user messages", () => {
    const result = convertToInflectionChatMessages([
      {
        role: "user",
        content: [{ type: "text", text: "" }],
      },
    ]);

    expect(result).toEqual([]);
  });
});

describe("system messages", () => {
  it("should convert system messages to instruction type", () => {
    const result = convertToInflectionChatMessages([
      {
        role: "system",
        content: "Be helpful and concise.",
      },
    ]);

    expect(result).toMatchSnapshot();
  });
});

describe("tool calls", () => {
  it("should stringify arguments in tool calls", () => {
    const result = convertToInflectionChatMessages(
      [
        {
          role: "assistant",
          content: [
            {
              type: "tool-call",
              args: { key: "arg-value", nested: { value: 123 } },
              toolCallId: "tool-call-id-1",
              toolName: "tool-1",
            },
          ],
        },
      ],
      "inflection_3_with_tools"
    );

    expect(result).toMatchSnapshot();
  });

  it("should stringify results in tool responses", () => {
    const result = convertToInflectionChatMessages(
      [
        {
          role: "tool",
          content: [
            {
              type: "tool-result",
              toolCallId: "tool-call-id-1",
              toolName: "tool-1",
              result: { key: "result-value", nested: { value: 123 } },
            },
          ],
        },
      ],
      "inflection_3_with_tools"
    );

    expect(result).toMatchSnapshot();
  });

  it("should handle mixed text and tool calls in one message", () => {
    const result = convertToInflectionChatMessages(
      [
        {
          role: "assistant",
          content: [
            { type: "text", text: "Let me check that." },
            {
              type: "tool-call",
              args: { key: "arg-value" },
              toolCallId: "tool-call-id-1",
              toolName: "tool-1",
            },
          ],
        },
      ],
      "inflection_3_with_tools"
    );

    expect(result).toMatchSnapshot();
  });

  it("should handle multiple tool calls in sequence", () => {
    const result = convertToInflectionChatMessages(
      [
        {
          role: "assistant",
          content: [
            {
              type: "tool-call",
              args: { key: "first-value" },
              toolCallId: "call-1",
              toolName: "tool-1",
            },
          ],
        },
        {
          role: "tool",
          content: [
            {
              type: "tool-result",
              toolCallId: "call-1",
              toolName: "tool-1",
              result: { key: "first-result" },
            },
          ],
        },
        {
          role: "assistant",
          content: [
            {
              type: "tool-call",
              args: { key: "second-value" },
              toolCallId: "call-2",
              toolName: "tool-2",
            },
          ],
        },
        {
          role: "tool",
          content: [
            {
              type: "tool-result",
              toolCallId: "call-2",
              toolName: "tool-2",
              result: { key: "second-result" },
            },
          ],
        },
      ],
      "inflection_3_with_tools"
    );

    expect(result).toMatchSnapshot();
  });

  it("should throw error for tool calls with non-tool models", () => {
    expect(() =>
      convertToInflectionChatMessages(
        [
          {
            role: "assistant",
            content: [
              {
                type: "tool-call",
                args: { key: "value" },
                toolCallId: "id-1",
                toolName: "tool-1",
              },
            ],
          },
        ],
        "inflection_3_pi"
      )
    ).toThrow(UnsupportedFunctionalityError);
  });
});

describe("assistant messages", () => {
  it("should convert basic assistant messages", () => {
    const result = convertToInflectionChatMessages([
      {
        role: "assistant",
        content: [{ type: "text", text: "Hello!" }],
      },
    ]);

    expect(result).toMatchSnapshot();
  });

  it("should add prefix true to trailing assistant messages", () => {
    const result = convertToInflectionChatMessages([
      {
        role: "user",
        content: [{ type: "text", text: "Hello" }],
      },
      {
        role: "assistant",
        content: [{ type: "text", text: "Hello!" }],
      },
    ]);

    expect(result).toMatchSnapshot();
  });
});
