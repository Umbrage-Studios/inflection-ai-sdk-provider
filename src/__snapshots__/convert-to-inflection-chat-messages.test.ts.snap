// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`assistant messages > should add prefix true to trailing assistant messages 1`] = `
[
  {
    "text": "Hello",
    "type": "Human",
  },
  {
    "text": "Hello!",
    "type": "AI",
  },
]
`;

exports[`assistant messages > should convert basic assistant messages 1`] = `
[
  {
    "text": "Hello!",
    "type": "AI",
  },
]
`;

exports[`system messages > should convert system messages to instruction type 1`] = `
[
  {
    "text": "Be helpful and concise.",
    "type": "Instruction",
  },
]
`;

exports[`tool calls > should handle mixed text and tool calls in one message 1`] = `
[
  {
    "text": "Let me check that.",
    "tool_calls": [
      {
        "function": {
          "arguments": "{"key":"arg-value"}",
          "name": "tool-1",
        },
        "id": "tool-call-id-1",
        "type": "function",
      },
    ],
    "type": "AI",
  },
]
`;

exports[`tool calls > should handle multiple tool calls in sequence 1`] = `
[
  {
    "text": "",
    "tool_calls": [
      {
        "function": {
          "arguments": "{"key":"first-value"}",
          "name": "tool-1",
        },
        "id": "call-1",
        "type": "function",
      },
    ],
    "type": "AI",
  },
  {
    "text": "{"key":"first-result"}",
    "tool_call_id": "call-1",
    "type": "Tool",
  },
  {
    "text": "",
    "tool_calls": [
      {
        "function": {
          "arguments": "{"key":"second-value"}",
          "name": "tool-2",
        },
        "id": "call-2",
        "type": "function",
      },
    ],
    "type": "AI",
  },
  {
    "text": "{"key":"second-result"}",
    "tool_call_id": "call-2",
    "type": "Tool",
  },
]
`;

exports[`tool calls > should stringify arguments in tool calls 1`] = `
[
  {
    "text": "",
    "tool_calls": [
      {
        "function": {
          "arguments": "{"key":"arg-value","nested":{"value":123}}",
          "name": "tool-1",
        },
        "id": "tool-call-id-1",
        "type": "function",
      },
    ],
    "type": "AI",
  },
]
`;

exports[`tool calls > should stringify results in tool responses 1`] = `
[
  {
    "text": "{"key":"result-value","nested":{"value":123}}",
    "tool_call_id": "tool-call-id-1",
    "type": "Tool",
  },
]
`;

exports[`user messages > should combine multiple text parts in user messages 1`] = `
[
  {
    "text": "Hello world!",
    "type": "Human",
  },
]
`;

exports[`user messages > should convert basic user messages 1`] = `
[
  {
    "text": "Hello",
    "type": "Human",
  },
]
`;
